README.md
# CertRenewal

Files used for cert renewals in AD and AD LDS


CHANGES

** Change History

# General Changes
    v1.1.2
    - Updated Instructions with additional detailed steps for clarification

Set-CertINF.ps1
    v1.0.0
    - Created function to generate cert.inf file

Get-CSR.ps1
    v1.0.0
    - Created CSR cmdlet

Get-Cert.ps1
    v1.1.2
    - Added parameter to enable Certificate Transparency options

    v1.1.0
    - Updated script to do ONLY CertMan function
    - Simplified name Get-SSLCert to Get-Cert

    v1.0.6
    - Fixed error with function=exit

    v1.0.0
    - Created Initial script to generate certificates via certman

Install-Cert.ps1
    v1.1.0
    - Simplified name for easier usage Install-SSLCert to Install-Cert

    v1.0.0
    - Created function to install generated certificate on a target system
    - Added CertName option for certs that are named differently than the target computer
    - Updated logic for local usage

Register-DCCert.ps1 (deprecated)
    v1.0.6
    - Fixed error with function=exit

    v1.0.5
    - Update Write-Host to Write-Output
    - Converted to function
    - Configure Client_rp via switch to match certman chosen
    - Updated credential prompts with descriptive text
    - Forced Get-Credential to command-prompt to handle certman style creds

    v1.0.0
    - Created script (adapted from Tom Hartsaw script)

Install-ServiceCert.psm1
    v1.0.2
    - Fixed Write-Output command
    - Added cert permissions logic

    v1.0.0
    - Created initial script

Set-CertPerm.ps1
    v1.0.0
    - Created Initial script


Set-CertPerm.ps1

<#
Script: Set-CertPerm.psm1
Author: CJ Ramseyer
Date: 11/20/2018
Modified: 
Version: v1.0.-

BREAKING CHANGES:
OTHER CHANGES:
- Created Cerrt Permission script

DISCLAIMER: 2023 CJ Ramseyer, All rights reserved.
This sample script is not supported under any support program or service. The sample script is provided AS-IS without warranty of any kind.
CJ Ramseyer disclaims all implied warranties including, without limitation, any implied warranties of merchantability or of fitness for a particular purpose.
The entire risk arising out of the use or performance of the sample script remains with you.
In no event shall CJ Ramseyer, its authors, or anyone else involved in the creation, production, or delivery of the scripts be liable for any damages whatsoever
(including, without limitation, damages for loss of business profits, business interruption, loss of business information, or other pecuniary loss) arising out
of the use of or inability to use the sample script, even if CJ Ramseyer has been advised of the possibility of such damages. 
#>

#Requires -Version 4.0

<#
.Synopsis
    Script to set ACL on Certificate directory for NetWork Service account
.Description
    Script to set ACL on Certificate directory for NetWork Service account
.Example
    .\Set-CertPerm.ps1
.INPUTS
    Script has no inputs
.OUTPUTS
    Script has no outputs
.NOTES
    Script does not require any option.  Used only to give network service account permission to read Certificate directory
#>

 $fullpath = "C:\Documents and Settings\All Users\Application Data\Microsoft\Crypto\RSA\MachineKeys"
 $acl=Get-Acl -Path $fullPath
 $permission="NT AUTHORITY\NETWORK SERVICE","Read","Allow"
 $accessRule=new-object System.Security.AccessControl.FileSystemAccessRule $permission
 $acl.AddAccessRule($accessRule)
 Set-Acl $fullPath $acl
 

functions/Get-CSR.ps1

<#
Script: Get-CSR.ps1
Author: CJ Ramseyer
Date: 11/19/2018
Modified: 
Version: v1.0.0

BREAKING CHANGES:
OTHER CHANGES:
- Created CSR cmdlet

DISCLAIMER: 2023 CJ Ramseyer, All rights reserved.
This sample script is not supported under any support program or service. The sample script is provided AS-IS without warranty of any kind.
CJ Ramseyer disclaims all implied warranties including, without limitation, any implied warranties of merchantability or of fitness for a particular purpose.
The entire risk arising out of the use or performance of the sample script remains with you.
In no event shall CJ Ramseyer, its authors, or anyone else involved in the creation, production, or delivery of the scripts be liable for any damages whatsoever
(including, without limitation, damages for loss of business profits, business interruption, loss of business information, or other pecuniary loss) arising out
of the use of or inability to use the sample script, even if CJ Ramseyer has been advised of the possibility of such damages. 
#>

#Requires -Version 4.0

Function Get-CSR
{
    <#
    .Synopsis
        Function to generate CSR (Certificate Signing Request) from the provided cert.inf file
    .Description
        Function to generate CSR (Certificate Signing Request) from the provided cert.inf file
    .Parameter CertName
        Required parameter to specify the name of the cert to be used
    .Parameter CertINF
        Optional parameter to specify the INF file to be used
    .Parameter enhanced
        Optional parameter to add enhanced features to cert such as smartcardlogin
    .Parameter ComputerName
        Optional parameter to specify a remote target machine
    .Parameter PSExec
        Optional parameter to use PSExec instead of RemotePS
    .NOTES
        If RemotePS does not work, copy INF file to target machine and generate CSR manually via: certreq -q -f -new "c:\temp\<SourceINF>.inf" "c:\temp\<TargetCSR>.csr"
        When done, copy .csr file back to local machine and rerun command using -CertONLY Parameter
    #>

    [CmdletBinding()]
    param
    (
        [parameter(Mandatory=$true)][string]$CertName,
        [parameter(Mandatory=$false)][string]$CertInf = "$CertName.inf",
        [parameter(Mandatory=$false)][string]$CertPath = "C:\temp",
        [Parameter(Mandatory=$false,ValueFromPipeline=$true)][string[]]$ComputerName,
        [parameter(Mandatory=$false)][switch]$PSExec
    )

    Set-PSDebug -Strict

    if(!([string]::IsNullOrEmpty($ComputerName)))
    {
        if(Test-Path "$Certinf")
        {
            $Credential = Get-Credential -Message 'Enter Remote System Admin Credentials (i.e. devad\$$cramse22)'
            New-PSDrive -Name "K" -PSProvider FileSystem -Root "\\$ComputerName\c`$\temp" -Credential $Credential
            Copy-Item "$CertPath\$CertName.inf" -Destination "K:\$CertName.inf"
        }
        else
        {
            Write-Output "Certificate Information file not found to generate cert"
            Break
        }    
    }

    if(!([string]::IsNullOrEmpty($ComputerName)) -and !($PSExec))
    {
        if($Credential -ne [System.Management.Automation.PSCredential]::Empty)
        {
            invoke-command -computername $ComputerName -Credential $Credential -scriptblock {param($s) certreq -q -f -new "c:\temp\$s.inf" "c:\temp\$s.csr" 2>1} -ArgumentList $CertName
        }
        else
        {
            invoke-command -computername $ComputerName -scriptblock {param($s) certreq -q -f -new "c:\temp\$s.inf" "c:\temp\$s.csr" 2>1} -ArgumentList $CertName
        }

        while (!(test-Path "K:\$CertName.csr"))
        {
            wait-event -timeout 2
            Write-Output "Waiting for $ComputerName"
        }
        if(Test-Path "K:\$CertName.csr")
        {
            Copy-Item "K:\$CertName.csr" -Destination "C:\temp\$CertName.csr"
        }
        else
        {
            Write-Output "Could not find K:\$CertName.csr"
            Write-Output "If csr only needs signing, use -CertONLY switch;ensure that csr is in c:\temp first"
            Break
        }
        Remove-PSDrive -Name "K"
    }
    elseif(!([string]::IsNullOrEmpty($ComputerName)) -and ($PSExec))
    {
        #Use PSExec for the remote command instead of remote powershell
        $PSCMD = "certreq -q -f -new `"c:\temp\$CertName.inf`" `"c:\temp\$CertName.csr`""
        psexec \\$ComputerName -u "$($Credential.UserName)" -p "$($Credential.Password)" CMD /c "$PSCMD"
    }
    else
    {
        Invoke-Command -ScriptBlock {param($s) certreq -q -f -new "c:\temp\$s.inf" "c:\temp\$s.csr"} -ArgumentList $CertName
    }

    while (!(test-Path C:\temp\$CertName.csr))
    {
        wait-event -timeout 2
        Write-Output "Waiting for $ComputerName"
    }
}


functions/Get-Cert.ps1

<#
Script: Get-Cert.ps1
Author: CJ Ramseyer
Date: 9/24/2018
Modified: 8/16/2019
Version: v1.1.2

BREAKING CHANGES:
OTHER CHANGES:
- Added parameter to enable Certificate Transparency options

DISCLAIMER: 2023 CJ Ramseyer, All rights reserved.
This sample script is not supported under any support program or service. The sample script is provided AS-IS without warranty of any kind.
CJ Ramseyer disclaims all implied warranties including, without limitation, any implied warranties of merchantability or of fitness for a particular purpose.
The entire risk arising out of the use or performance of the sample script remains with you.
In no event shall CJ Ramseyer, its authors, or anyone else involved in the creation, production, or delivery of the scripts be liable for any damages whatsoever
(including, without limitation, damages for loss of business profits, business interruption, loss of business information, or other pecuniary loss) arising out
of the use of or inability to use the sample script, even if CJ Ramseyer has been advised of the possibility of such damages. 
#>

#Requires -Version 4.0

Function Get-Cert
{
    <#
    .Synopsis
        Function to generate SSL Certificate based on a provided cert.inf file
    .Description
        Function to generate SSL Certificate based on a provided cert.inf file
    .Parameter CertName
        Required parameter to specify the name of the cert to be used
    .Parameter ComputerName
        Optional parameter to specify a remote target machine
    .Parameter Type
        Optional parameter to change the cert type from the P7B default
    .Parameter enhanced
        Optional parameter to add enhanced features to cert such as smartcardlogin
    .Parameter CertMan
        Optional parameter to specify a differrent CertMan to use instead of the default (Prod)
    .Parameter BusinessOwner
        Optional parameter to change the default business owner
    .Parameter email
        Optional parameter to change the default businessowner email
    .Parameter RenewContact
        Optional parameter to specify the CDSID of the renewal contact
    .Parameter acr
        Optional parameter to specify the ACR number associated with the request
    .Parameter itms
        Optional parameter to specify the itms number associated with the request
    .Parameter CTEnable
        Optional parameter used to enable/disable Certificate Transparency (Option is YES or NO only)
    .Example
        Get-Cert -CertName ADLDSEngSand.company.com -Certman www.certman.company.com
        This would get a signed cert from production CertMan named adldsengsand.company.com and assign CJ Ramseyer as the cert owner without Certificate Transparency
    .NOTES
        CDSID must be a member of ADCS-Groups-SSL-API-Functions to access REST API
        Remote PowerShell MUST BE Configured AND Enabled
        If RemotePS does not work, copy INF file to target machine and generate CSR manually via: certreq -q -f -new "c:\temp\<SourceINF>.inf" "c:\temp\<TargetCSR>.csr"
        When done, copy .csr file back to local machine and rerun command using -CertONLY Parameter
    #>

    [CmdletBinding()]
    param
    (
        [parameter(Mandatory=$true)][string]$CertName,
        [Parameter(Mandatory=$false,ValueFromPipeline=$true)][string[]]$ComputerName,
        [parameter(Mandatory=$false)][ValidateSet("P7B","DER","PEM")][string]$Type = "P7B",
        [parameter(Mandatory=$false)][switch]$enhanced,
        [parameter(Mandatory=$false)][ValidateSet("www.certman.company.com","wwwqa.certman.company.com","wwwdev.certman.company.com")][string]$CertMan = "www.certman.company.com",
        [parameter(Mandatory=$false)][string]$BusinessOwner = "cjramseyer",
        [parameter(Mandatory=$false)][string]$email = "cjr351@gmail.com",
        [parameter(Mandatory=$false)][string]$RenewContact = "cjramseyer",
        [parameter(Mandatory=$false)][string]$acr = "2023-0001",
        [parameter(Mandatory=$false)][string]$itms = "00001",
        [parameter(Mandatory=$false)][switch]$PSExec,
        [parameter(Mandatory=$false)][ValidateSet("YES","NO")][string]$CTEnable = "NO"
    )

    Set-PSDebug -Strict

    Set-Variable -Name ServerAuthentication -Value 1.3.6.1.5.5.7.3.1 -Option Constant
    Set-Variable -Name ClientAuthentication -Value 1.3.6.1.5.5.7.3.2 -Option Constant
    Set-Variable -Name EmailProtection -Value 1.3.6.1.5.5.7.3.4 -Option Constant
    Set-Variable -Name IPSecEndSystem -Value 1.3.6.1.5.5.7.3.5 -Option Constant
    Set-Variable -Name IPSecTunnel -Value 1.3.6.1.5.5.7.3.6 -Option Constant
    Set-Variable -Name IPSecUser -Value 1.3.6.1.5.5.7.3.7 -Option Constant
    Set-Variable -Name KDCAuthentication -Value 1.3.6.1.5.2.3.5 -Option Constant
    Set-Variable -Name SmartCardLogon -Value 1.3.6.1.4.1.311.20.2.2 -Option Constant
    Set-Variable -Name IPSecIKE -Value 1.3.6.1.5.5.7.3.17 -Option Constant
    Set-Variable -Name IKEIntermediate -Value 1.3.6.1.5.5.8.2.2 -Option Constant
    Set-Variable -Name SANSupport -value 2.5.29.17 -Option Constant
    Set-Variable -Name ServerFacingUnkown -Value 0 -Option Constant
    Set-Variable -Name ServerFacingInternal -Value 1 -Option Constant
    Set-Variable -Name ServerFacingExternal -Value 2 -Option Constant
    Set-Variable -Name ServerFacingMixed -Value 3 -Option Constant
    Set-Variable -Name CTEnableUnkown -Value 0 -Option Constant
    Set-Variable -Name CTEnablYes -Value 1 -Option Constant
    Set-Variable -Name CTEnablNo -Value 2 -Option Constant

    $ServerFacing = $ServerFacingExternal
    if($CTEnable -eq "YES")
    {
        [int]$CTEnable = "$CTEnablYes"
    }
    else
    {
        [int]$CTEnable = "$CTEnablNo"
    }

    if(!$enhanced)
    {
        $enhancedusage = "$ServerAuthentication,$ClientAuthentication"
    }
    else
    {
        $enhancedusage = "$ServerAuthentication,$ClientAuthentication,$KDCAuthentication,$SmartCardLogon"
    }

    if(Test-Path "$env:homedrive\$env:homepath\certdefaults.ini")
    {
        Invoke-Expression $(Get-Content "$env:homedrive\$env:homepath\certdefaults.ini" | Out-String)
    }
    else
    {
        if((!$BusinessOwner) -or (!$acr) -or (!$itms) -or (!$RenewContact))
        {
            Write-Output "No defaults file found;if defaults needed create certdefaults.ini here: $env:homedrive\$env:homepath" | Out-File $ErrorLog -Append
            Write-Output "CAUTION: Default file overrides any value specified in both the defaults and the command line" | Out-File $ErrorLog -Append
            $Ask4File = Read-Host "Do you want to set default values?"
            if($Ask4File -eq "YES")
            {
                Out-File -FilePath "$env:homedrive\$env:homepath\certdefaults.ini"
                '$BusinessOwner = ' | Out-File -FilePath "$env:homedrive\$env:homepath\certdefaults.ini" -Append
                '$email = ' | Out-File -FilePath "$env:homedrive\$env:homepath\certdefaults.ini" -Append
                '$RenewContact = ' | Out-File -FilePath "$env:homedrive\$env:homepath\certdefaults.ini" -Append
                '$acr = ' | Out-File -FilePath "$env:homedrive\$env:homepath\certdefaults.ini" -Append
                '$itms = ' | Out-File -FilePath "$env:homedrive\$env:homepath\certdefaults.ini" -Append
                Write-Output "Please update certdefaults.ini and rerun script" | Out-File $ErrorLog -Append
                Break
            }
        }
    }

    Switch ($CertMan)
    {
        "www.certman.company.com"
        {[string]$Client_rp = "urn:certman:prod"}
        "wwwqa.certman.company.com"
        {[string]$Client_rp = "urn:certman:qa"}
        "wwwdev.certman.company.com"
        {[string]$Client_rp = "urn:certman:dev"}
        Default
        {[string]$Client_rp = "urn:certman:prod"}
    }

    # Get credential for REST API
    #Force Get-Credential to use command-line to capture proper UserName for
    $key = 'HKLM:\SOFTWARE\Microsoft\PowerShell\1\ShellIds'
    Set-ItemProperty $key ConsolePrompting True
    $Creds = Get-Credential -Message "Enter CertMan credentials (i.e. xna1\\<cdsid>)"
    $reqcdsid = $Creds.username
    $pwd = $Creds.GetNetworkCredential().password
    $Bytes = [System.Text.Encoding]::ASCII.GetBytes($pwd)
    $Cred64= [Convert]::ToBase64String($bytes)
    $jsonat ='{"client_id": "' + $reqcdsid + '","client_pwdb64": "'+ $Cred64 + '","client_rp": "' + $Client_rp + '"}'
    $url = 'https://' + $Certman + '/REST/api/v1/oauth/jwttoken'
    $responseat = Invoke-RestMethod $url -ContentType 'application/json' -Method post -Body $jsonat
    if($responseat.Status -ne "Success")
    {
        Write-Output "Certman Authentication failed!"
        Remove-Variable -Name * -ErrorAction SilentlyContinue
        Break
    }
    else
    {
        Write-Output "CertMan Authentication successful!"
    }
    $access_token = $responseat.access_token
    #Remove registry entry force Get-Credential to command line
    Remove-ItemProperty $key ConsolePrompting

    $key = "$CertName"
    $sans = '"'+$key.Substring($key.IndexOf(".") +1)+'"'

    if ([string]$csr = Get-Content -Path "C:\temp\$CertName.csr")
    {
        # Query certman to determine if this is a new cert or renewall

        $headers = @{"Auth"="Bearer $access_token"}
        #$query_string = "AnyOwner=THARTSAW&MaxRecordCount=10"
        $query_string = "CN=$CertName&MaxRecordCount=1"
        $url = "https://" + $Certman + "/REST/api/v1/SSLCertificateRequests?$query_string"
        $responseq = Invoke-RestMethod $url -Method get -Headers $headers 
        #$responseq | ConvertTo-Json 

        if ($responseq.TotalRecordCount -gt 0)
        {
            #RENEW 
            $reqnum = $responseq.Certificates.Item(0).RequestNumber  
            $json2 = 
            @"
            {  
                "AcrNumber": "$acr",  
                "BusinessOwner": "$BusinessOwner",  
                "CertificateTypeRequested" : "$Type", 
                "CertificateSigningRequest": "$csr",
                "Email": "$email", 
                "EnhancedKeyUsage": "$enhancedusage", 
                "ExpireNotificationFlag": true,  
                "ITMS": "$itms",  
                "PAHAttestation": true, 
                "RenewContact" : "$RenewContact", 
                "SAN": [$sans],
                "ServerFacing": $ServerFacing, 
                "CTEnable": $CTEnable, 
                "CTEnableAttestation": true
            }
"@

            $headers = @{"Auth"="Bearer $access_token"}
            $url = "https://" + $Certman + "/REST/api/v1/SSLCertificateRequests/$reqnum"
            $response = Invoke-RestMethod $url -ContentType 'application/json' -Method post -Headers $headers -Body $json2
            #$response | ConvertTo-Json 
            $response.status
            $response.message
            $response.ssl_certificate.certStream |Out-File -FilePath C:\temp\$key.p7b -Encoding ascii
        }
        else
        {
            #NEW
            $json2 = 
            @"
            {
                "AcrNumber": "$acr",  
                "Attestor": "$reqcdsid",  
                "BusinessOwner": "$BusinessOwner",  
                "CertificateTypeRequested" : "$Type", 
                "CN": "$key", 
                "CertificateSigningRequest": "$csr",
                "Email": "$email", 
                "EnhancedKeyUsage": "$enhancedusage",  
                "ExpireNotificationFlag": true,  
                "ITMS": "$itms",  
                "PAHAttestation": true, 
                "RenewContact" : "$RenewContact", 
                "SAN": [$sans],
                "ServerFacing": $ServerFacing, 
                "CTEnable": $CTEnable, 
                "CTEnableAttestation": true
            }
"@

            $headers = @{"Auth"="Bearer $access_token"}
            $url = "https://" + $certman + "/REST/api/v1/SSLCertificateRequests"
            $response = Invoke-RestMethod $url -ContentType 'application/json' -Method post -Headers $headers -Body $json2
            #$response | ConvertTo-Json 
            $response.status
            $response.message
            if($response.status -eq "Success")
            {
                $response.ssl_certificate.certStream |Out-File -FilePath C:\temp\$CertName.p7b -Encoding ascii
            }
        }
    }
    else 
    {
        Write-Output "$CertName.csr not found"
    }
}


functions/Set-CertINF.ps1

<#
Script: Set-CertINF.ps1
Author: CJ Ramseyer
Date: 9/24/2018
Modified: 
Version: v1.0.0

BREAKING CHANGES:
OTHER CHANGES:
- Created function to generate cert.inf file

DISCLAIMER: 2023 CJ Ramseyer, All rights reserved.
This sample script is not supported under any support program or service. The sample script is provided AS-IS without warranty of any kind.
CJ Ramseyer disclaims all implied warranties including, without limitation, any implied warranties of merchantability or of fitness for a particular purpose.
The entire risk arising out of the use or performance of the sample script remains with you.
In no event shall CJ Ramseyer, its authors, or anyone else involved in the creation, production, or delivery of the scripts be liable for any damages whatsoever
(including, without limitation, damages for loss of business profits, business interruption, loss of business information, or other pecuniary loss) arising out
of the use of or inability to use the sample script, even if CJ Ramseyer has been advised of the possibility of such damages. 
#>

#Requires -Version 4.0

Function Set-CERTINF
{
    <#
    .Synopsis
        Function to generate the Certtificate information file used to generate the certificate request file
    .Description
        Function to generate the Certtificate information file used to generate the certificate request file
    .Parameter CertName
        Required parameter to specify the name of the certificate
    .Parameter KeyLength
        Optional parameter to specify a differnet key length
        NOTE: ONLY Values greater than 2048 whole integers will be accepted
    .Parameter enhanced
        Optional parameter ro enable extended cert attriobutes such as smartcardlogin
    .Parameter sans
        Optional parameter to specify a comma delimited list of subject alternative names
        i.e. "fdsdev.ford.com","fdsdev.ford.com"
    .Example
        Set-CertINF -Name fdsdev.ford.com -sans "fdsdev.ford.com","fdsdevs.ford.com","fmc128025.fmcc.ford.com","fmc128028.fmcc.ford.com"
    #>

    [CmdletBinding()]
    param
    (
        [Parameter(Mandatory=$true,ValueFromPipeline=$true)][string[]]$CertName,
        [parameter(Mandatory=$false)][ValidateRange(2048,[int]::MaxValue)][int16]$KeyLength = 2048,
        [parameter(Mandatory=$false)][switch]$enhanced,
        [parameter(Mandatory=$false)][string[]]$sans
    )

    Set-PSDebug -Strict

    Set-Variable -Name ServerAuthentication -Value 1.3.6.1.5.5.7.3.1 -Option Constant
    Set-Variable -Name ClientAuthentication -Value 1.3.6.1.5.5.7.3.2 -Option Constant
    Set-Variable -Name EmailProtection -Value 1.3.6.1.5.5.7.3.4 -Option Constant
    Set-Variable -Name IPSecEndSystem -Value 1.3.6.1.5.5.7.3.5 -Option Constant
    Set-Variable -Name IPSecTunnel -Value 1.3.6.1.5.5.7.3.6 -Option Constant
    Set-Variable -Name IPSecUser -Value 1.3.6.1.5.5.7.3.7 -Option Constant
    Set-Variable -Name KDCAuthentication -Value 1.3.6.1.5.2.3.5 -Option Constant
    Set-Variable -Name SmartCardLogon -Value 1.3.6.1.4.1.311.20.2.2 -Option Constant
    Set-Variable -Name IPSecIKE -Value 1.3.6.1.5.5.7.3.17 -Option Constant
    Set-Variable -Name IKEIntermediate -Value 1.3.6.1.5.5.8.2.2 -Option Constant
    Set-Variable -Name SANSupport -value 2.5.29.17 -Option Constant
    Set-Variable -Name ServerFacingUnkown -Value 0 -Option Constant
    Set-Variable -Name ServerFacingInternal -Value 1 -Option Constant
    Set-Variable -Name ServerFacingExternal -Value 2 -Option Constant
    Set-Variable -Name ServerFacingMixed -Value 3 -Option Constant
    Set-Variable -Name CTEnableUnkown -Value 0 -Option Constant
    Set-Variable -Name CTEnablYes -Value 1 -Option Constant
    Set-Variable -Name CTEnablNo -Value 2 -Option Constant

    $Signature = '"$Windows NT$"'

    Write-Output "Processing $ComputerName"

    $key="$CertName"

    $INF = @()
    $INF += '[Version]'
    $INF += 'Signature='+"$Signature"
    $INF += '[NewRequest]'
    $INF += ''
    $INF += 'Subject='+"`"CN=$key, O=Big LDAP Company, OU=IT, L=Somecity, ST=Somestate, C=US`""
    $INF += ''
    $INF += 'KeySpec = 1'
    $INF += 'KeyLength = '+"$KeyLength"
    $INF += 'Exportable = TRUE'
    $INF += 'MachineKeySet = TRUE'
    $INF += ''
    $INF += 'SMIME = False'
    $INF += 'PrivateKeyArchive = FALSE'
    $INF += 'UserProtected = FALSE'
    $INF += 'UseExistingKeySet = FALSE'
    $INF += 'ProviderName = "Microsoft RSA SChannel Cryptographic Provider"'
    $INF += 'ProviderType = 12'
    $INF += 'RequestType = PKCS10'
    $INF += 'KeyUsage = 0xa0'

    if($enhanced)
    {
        $INF += ''
        $INF += '[EnhancedKeyUsageExtension]'
        $INF += 'OID='+"$ServerAuthentication"
        $INF += 'OID='+"$ClientAuthentication"
        $INF += 'OID='+"$SmartCardLogon"
        $INF += 'OID='+"$KDCAuthentication"
    }
    elseif((!$enhanced) -and (!$sans))
    {
        $INF += ''
        $INF += '[EnhancedKeyUsageExtension]'
        $INF += 'OID='+"$ServerAuthentication"
        $INF += 'OID='+"$ClientAuthentication"
    }

    if($sans)
    {
        $INF += ''
        $INF += '[Extensions]'
        $INF += ''
        $INF += "$SANSupport"+' = "{text}"'
        foreach($san in $sans)
        {
            $INF += '_continue_ = '+"`"dns=$san&`""
        }
    }

    $sans = '"'+$key.Substring($key.IndexOf(".") +1)+'"'

    Write-Output "Certificate Information file is being generated `n "

    $INF | out-file -Filepath "c:\temp\$CertName.inf" -Encoding ascii
}


functions/Install-Cert.ps1

<#
Script: Install-SSLCert.ps1
Author: CJ Ramseyer
Date: 9/24/2018
Modified: 11/20/2018
Version: v1.1.0

BREAKING CHANGES:
OTHER CHANGES:
- Simplified name for easier usage Install-SSLCert to Install-Cert

DISCLAIMER: 2023 CJ Ramseyer, All rights reserved.
This sample script is not supported under any support program or service. The sample script is provided AS-IS without warranty of any kind.
CJ Ramseyer disclaims all implied warranties including, without limitation, any implied warranties of merchantability or of fitness for a particular purpose.
The entire risk arising out of the use or performance of the sample script remains with you.
In no event shall CJ Ramseyer, its authors, or anyone else involved in the creation, production, or delivery of the scripts be liable for any damages whatsoever
(including, without limitation, damages for loss of business profits, business interruption, loss of business information, or other pecuniary loss) arising out
of the use of or inability to use the sample script, even if CJ Ramseyer has been advised of the possibility of such damages. 
#>

#Requires -Version 4.0

Function Install-Cert
{
    <#
    .Synopsis
        Function to install a generated cert to a target system under machine or a service
    .Description
        Function to install a generated cert to a target system under machine or a service
    .Parameter CertName
        Required parameter to specify the name of the certificate to be installed
    .Parameter ComputerName
        Optional parameter to install the certificate on a remote machine
    .Parameter InstallMode
        Optional parameter to specify the install mode for the certificate
        Options are Machine or Service.
        Currently ONLY Machine is supported in this script
    .Parameter ServiceName
        Optional Parameter to specify the name of the service where the certificate needs to be installed
    .NOTES

    #>

    [CmdletBinding()]
    param
    (
        [parameter(Mandatory=$true)][string]$CertName,
        [Parameter(Mandatory=$false,ValueFromPipeline=$true)][string]$ComputerName,
        [parameter(Mandatory=$false)][ValidateSet("Machine","Service")][string]$InstallMode = "Machine",
        [Parameter(Mandatory=$false,ValueFromPipelineByPropertyName=$true)][string]$ServiceName
    )

    switch ($InstallMode)
    {
        "Machine"
        {
            if(!([string]::IsNullOrEmpty($ComputerName)))
            {
                $Credential = Get-Credential
                New-PSDrive -Name "K" -PSProvider FileSystem -Root "\\$ComputerName\c`$\temp" -Credential $Credential
                Copy-Item C:\temp\$CertName.p7b -destination "K:\$CertName.p7b" -force
                while (!(test-Path "K:\$CertName.p7b"))
                {
                    Wait-Event -timeout 2
                    Write-Output "Waiting for $ComputerName"
                }

                if($Credential -ne [System.Management.Automation.PSCredential]::Empty)
                {
                    invoke-command -computername $ComputerName -Credential $Credential -scriptblock {param($s) certreq -q -accept c:\temp\$s.p7b 2>1} -ArgumentList $CertName
                }
                else
                {
                    invoke-command -computername $ComputerName -scriptblock {param($s) certreq -q -accept c:\temp\$s.p7b 2>1} -ArgumentList $CertName
                }
            }
            else
            {
                invoke-command -ScriptBlock {param($s) certreq -q -accept c:\temp\$s.p7b 2>1} -ArgumentList $CertName
            }
        }
        "Service"
        {
            Write-Output "Use Install-ServiceCert.psm1 locally on target system to install service based certificate;exiting"
        }
    }
}
